src/Domain/BlockLogicalCoordinates.cpp

```diff
@@ -0,0 +1,102 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "BlockLogicalCoordinates.hpp"
+
+#include <cstddef>
+#include <vector>
+
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Domain/Domain.hpp"  // IWYU pragma: keep
+#include "ErrorHandling/Error.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Literals.hpp"
+
+template <size_t Dim>
+void BlockLogicalCoordHolder<Dim>::pup(PUP::er& p) noexcept {
+  p | block_id;
+  p | block_logical_coords;
+}
+
+template <size_t Dim>
+bool operator==(const BlockLogicalCoordHolder<Dim>& lhs,
+                const BlockLogicalCoordHolder<Dim>& rhs) noexcept {
+  return lhs.block_id == rhs.block_id and
+         lhs.block_logical_coords == rhs.block_logical_coords;
+}
+template <size_t Dim>
+bool operator!=(const BlockLogicalCoordHolder<Dim>& lhs,
+                const BlockLogicalCoordHolder<Dim>& rhs) noexcept {
+  return not(lhs == rhs);
+}
+
+template <size_t Dim, typename Frame>
+std::vector<BlockLogicalCoordHolder<Dim>> block_logical_coordinates(
+    const Domain<Dim, Frame>& domain,
+    const tnsr::I<DataVector, Dim, Frame>& x) noexcept {
+  const size_t num_pts = get<0>(x).size();
+  std::vector<BlockLogicalCoordHolder<Dim>> block_logical_coords(num_pts);
+  for (size_t s = 0; s < num_pts; ++s) {
+    tnsr::I<double, Dim, Frame> x_global(0.0);
```

Comment:
Might we want a method to break tensors of datavectors into tensors of doubles? It doesn't seem unlikely that this could be used in other places

Reply:
I agree that this is awkward...